core/src/main/java/tachyon/worker/eviction/EvictLFU.java

```diff
@@ -0,0 +1,120 @@
+
+import com.google.common.collect.HashMultimap;
+
+import tachyon.Pair;
+import tachyon.worker.hierarchy.BlockInfo;
+import tachyon.worker.hierarchy.StorageDir;
+
+/**
+ * Used to evict old blocks among several StorageDirs by LFU strategy.
+ */
+public final class EvictLFU extends EvictLFUBase {
+
+  public EvictLFU(boolean lastTier) {
+    super(lastTier);
+  }
+
+  @Override
+  public synchronized Pair<StorageDir, List<BlockInfo>> getDirCandidate(StorageDir[] storageDirs,
+      Set<Integer> pinList, long requestBytes) {
+    /** blocks to be evicted */
+    List<BlockInfo> blockInfoList = new ArrayList<BlockInfo>();
+    /** current least frequency information of each StorageDir, except the evicted blocks */
+    Map<StorageDir, Pair<Long, Long>> dir2LFUBlocks = new HashMap<StorageDir, Pair<Long, Long>>();
+    /** blocks that have been selected to be evicted of each StorageDir */
+    HashMultimap<StorageDir, Long> dir2BlocksToEvict = HashMultimap.create();
+    /** the total size of evicted blocks of each StorageDir */
+    Map<StorageDir, Long> sizeToEvict = new HashMap<StorageDir, Long>();
+    // If no StorageDir has enough space for the request size, continue; if no block can be evicted,
+    // return null; and if eviction size plus free space of some StorageDir is larger than request
+    // size, return the Pair of StorageDir and blockInfoList
+    while (true) {
+      // Get least frequent block in StorageDir candidates
+      Pair<StorageDir, Long> candidate =
+          getLFUBlockCandidate(storageDirs, dir2LFUBlocks, dir2BlocksToEvict, pinList);
+      StorageDir dir = candidate.getFirst();
+      if (dir == null) {
+        return null;
+      }
+      long blockId = candidate.getSecond();
+      long blockSize = dir.getBlockSize(blockId);
+      // Add info of the block to the list
+      blockInfoList.add(new BlockInfo(dir, blockId, blockSize));
+      dir2BlocksToEvict.put(dir, blockId);
+      dir2LFUBlocks.remove(dir);
+      // Update eviction size for this StorageDir
+      long evictBytes;
+      if (sizeToEvict.containsKey(dir)) {
+        evictBytes = sizeToEvict.get(dir) + blockSize;
+      } else {
+        evictBytes = blockSize;
+      }
+      sizeToEvict.put(dir, evictBytes);
+      if (evictBytes + dir.getAvailableBytes() >= requestBytes) {
+        return new Pair<StorageDir, List<BlockInfo>>(dir, blockInfoList);
+      }
+    }
+  }
+
+  /**
+   * Get block to be evicted by choosing the least frequent block in StorageDir candidates
+   * 
+   * @param storageDirs StorageDir candidates that the space will be allocated in
+   * @param dir2LFUBlocks the least frequency information of each StorageDir
+   * @param dir2BlocksToEvict Ids of blocks that have been selected to be evicted
+   * @param pinList list of pinned files
+   * @return pair of StorageDir that contains the block to be evicted and Id of the block
+   */
+  private Pair<StorageDir, Long> getLFUBlockCandidate(StorageDir[] storageDirs,
```

Comment:
In both cases, the space needs to be evicted from one storage dir, so having a method to get the list of blocks that needs to be evicted from a storage dir to make X space would remove the need for the while loop and each storage dir block list only needs to be iterated once.

For evictLFU, the min of the max LFU in each storagedir's evict block list can be taken and all blocks with LFU less than it can be evicted.

However its a micro optimization and maybe the better solution is just to generate a min heap for each block set when we need to evict.